sydent/db/threepid_associations.py

```diff
@@ -227,22 +227,48 @@ def removeAssociation(self, medium, address):
         )
         self.sydent.db.commit()
 
-    def retrieveMxidFromHash(self, lookup_hash):
-        """Returns an mxid from a given lookup_hash value
+    def retrieveMxidsForHashes(self, addresses):
+        """Returns a mapping from hash: mxid from a list of given lookup_hash values
 
-        :param input_hash: The lookup_hash value to lookup in the database
-        :type input_hash: str
+        :param addresses: An array of lookup_hash values to check against the db
+        :type addresses: list[str]
 
-        :returns the mxid relating to the lookup_hash value if found,
-                 otherwise None
-        :rtype: str|None
+        :returns a dictionary of lookup_hash values to mxids of all discovered matches
+        :rtype: dict[str, str]
         """
         cur = self.sydent.db.cursor()
 
-        res = cur.execute(
-            "SELECT mxid FROM global_threepid_associations WHERE lookup_hash = ?", (lookup_hash,)
-        )
-        row = res.fetchone()
-        if not row:
-            return None
-        return row[0]
+        cur.execute("CREATE TEMPORARY TABLE tmp_retrieve_mxids_for_hashes ("
+                    "lookup_hash VARCHAR, "
+                    "mxid VARCHAR"
+                    ")")
+        cur.execute("CREATE INDEX tmp_retrieve_mxids_for_hashes_lookup_hash ON "
+                    "tmp_retrieve_mxids_for_hashes(lookup_hash)")
+
+        results = {}
+        try:
+            # Convert list of addresses to list of tuples of addresses
+            addresses = [(x,) for x in addresses]
+
+            inserted_cap = 0
+            while inserted_cap < len(addresses):
+                cur.executemany(
+                    "INSERT INTO tmp_retrieve_mxids_for_hashes(lookup_hash) "
+                    "VALUES (?)",
+                    addresses[inserted_cap:inserted_cap + 500]
+                )
+                inserted_cap += 500
+
+            res = cur.execute(
+                "SELECT gta.lookup_hash, gta.mxid FROM global_threepid_associations gta "
+                "JOIN tmp_retrieve_mxids_for_hashes "
+                "ON gta.lookup_hash = tmp_retrieve_mxids_for_hashes.lookup_hash"
```

Comment:
We still need to restrict and sort by timestamp as I don't think we can rely on having only one binding in the table for a hash, and we probably ought to continue honouring the expiry even though we never use it in practice.